enhancements/ingress/global-options-enable-hsts.md

```diff
@@ -0,0 +1,457 @@
+`reencrypt` routes that match the `haproxy.router.openshift.io/hsts_header` annotation:
+```gotemplate
+{{- /* hsts header in response: */}}
+ {{- $hstsOptionalTokenPattern := `(?:includeSubDomains|preload)` }}
+ {{- $hstsPattern := printf `(?:%[1]s[;])*max-age=(?:\d+|"\d+")(?:[;]%[1]s)*`  $hstsOptionalTokenPattern -}}
+...
+{{- if matchValues (print $cfg.TLSTermination) "edge" "reencrypt" }}
+    {{- with $hsts := firstMatch $hstsPattern (index $cfg.Annotations "haproxy.router.openshift.io/hsts_header") }}
+  http-response set-header Strict-Transport-Security {{$hsts}}
+    {{- end }}{{/* hsts header */}}
+  {{- end }}{{/* is "edge" or "reencrypt" */}}
+```
+
+#### Enable HSTS for all routes globally, based on the cluster's ingress domain
+To add the header to all HTTP responses from an `edge` or `reencrypt` route, add a new router environment
+variable called `ROUTER_DEFAULT_HSTS`.
+
+Add a new type `IngressControllerHSTSHeaderPolicy` to the API in
+`openshift/api/blob/master/operator/v1/types_ingress.go`, to capture the configuration of the HSTS Policy.
+```go
+// These are strings instead of enums so that they can be passed in easily from the HAProxy template
+const (
+    AllHSTSHeaderPolicyScope         = "All"
+    LimitedHSTSHeaderPolicyScope     = "Limited"
+)
+
+// HSTSDirectiveType is an optional directive that directs HSTS Policy configuration.
+//+kubebuilder:validation:Enum=preload;includeSubDomains
+type HSTSDirectiveType string
+const(
+    // PreloadHSTSDirective directs the client to include this host in its host preload list so that
+    // it never needs to do an initial load to get the HSTS header (note that this is not defined
+    // in RFC 6797 and is therefore client implementation-dependent).
+    PreloadHSTSDirective             HSTSDirectiveType = "preload"
+
+    // IncludeSubDomainsHSTSDirective means the HSTS Policy applies to any subdomains of the host's domain name
+    IncludeSubDomainsHSTSDirective   HSTSDirectiveType = "includeSubDomains"
+)
+
+type IngressControllerHSTSHeaderPolicy {
+	// maxAgeSeconds is the delta time range in seconds during which the host is regarded as an HSTS host.
+	// If set to 0, it negates the effect and the host is no longer regarded as an HSTS host.
+	// maxAgeSeconds is a time-to-live value, and if it is not refreshed, the HSTS policy for this host
+	// will eventually expire.
+	// +required
+    MaxAgeSeconds       int32 `json:"maxAgeSeconds"`
+
+    // HSTSDirectives is a list of optional HSTS directives that direct HSTS Policy configuration.
+    HSTSDirectives      []HSTSDirectiveType `json:"hstsDirectives,omitempty"`
+
+    // scope indicates whether this header policy is enabled for routes that have hosts with domains
+    // in the list specified in inScopeDomains (limited scope) or for all routes that have hosts under the
+    // cluster ingress domain (all scope).  If global HSTS is desired, set the scope field to "all".  The
```

Comment:
Would `All` scope enable the given HSTS settings for all of the routes served by the given Ingress Controller?

Reply:
Yes, it would add the HSTS header to all http responses for edge and reencrypt routes under the cluster ingress domain.  I added some more to clarify:

If HSTS is desired for all routes, set the scope field to "All". If HSTS is desired for only a limited set of routes, set the scope field to "Limited" and add a list of subdomains to the InScopeDomains field.

Reply:
> Yes, it would add the HSTS header to all http responses for edge and reencrypt routes under the cluster ingress domain. I added some more to clarify:

What if an Ingress Controller is serving routes with hosts that aren't under the cluster's ingress domain? Would `all` exclude those routes?

Reply:
No, if scope is "All" then all routes will get the configured HSTS Policy header.  I re-wrote this, ptal.

Reply:
It still says "for all routes that have hosts under the cluster ingress domain (all scope)".  Also, the second goal is, "Enable HSTS for all routes globally, based on the cluster's ingress domain".  I think in both cases that should be "all TLS routes" without any reference to the cluster's ingress domain, right?

Reply:
Also, the heading "Enable HSTS for all routes globally, based on the cluster's ingress domain" should omit "based on the cluster's ingress domain" if "All" means includes hosts that aren't under the ingress domain.

Reply:
Fixed.